[WTF][GLib] Rework FileSystem::openFile()

ntrrgc opened the pull request:
#### 78d2561b2bf73ecfc596b303f3321e6789c1f659
<pre>
[WTF][GLib] Rework FileSystem::openFile()
<a href="https://bugs.webkit.org/show_bug.cgi?id=254813">https://bugs.webkit.org/show_bug.cgi?id=254813</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

Fixes REGRESSION(259689@main). The previous implementation failed when
asked to create a file in ReadWrite mode.

This patch also increases the test coverage for FileSystem::openFile().
Previously, few argument combinations where tested, which caused the
regression to go unnoticed.

Read requests no longer create a read-write stream.

To accomodate that, PlatformFileHandle in GLib is now the generic
GSeekable* rather than GFileIOStream* (which implies a read-write file
handle).

This patch also introduces WTF::FileSystem::posixFileDescriptor() for
non-Windows platforms, which allows decoupling of the GIO API in a few
places in WebKit.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::genericGIOFileClose):
(WTF::FileSystemImpl::genericGIOFileQueryInfo):
(WTF::FileSystemImpl::genericGIOGetInputStream):
(WTF::FileSystemImpl::genericGIOGetOutputStream):
(WTF::FileSystemImpl::genericGIOGetFileDescriptorBased):
(WTF::FileSystemImpl::posixFileDescriptor):
(WTF::FileSystemImpl::fileSize):
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::openTemporaryFile):
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::closeFile):
(WTF::FileSystemImpl::seekFile):
(WTF::FileSystemImpl::truncateFile):
(WTF::FileSystemImpl::flushFile):
(WTF::FileSystemImpl::writeToFile):
(WTF::FileSystemImpl::readFromFile):
(WTF::FileSystemImpl::lockFile):
(WTF::FileSystemImpl::unlockFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::posixFileDescriptor):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/263367@main">https://commits.webkit.org/263367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b3d5f4af3e2d5a17427b6d283d3feeb82e2a58e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5837 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4577 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4445 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/4804 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3924 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5837 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2069 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3901 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/7302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3624 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3899 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/3968 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4146 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4378 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3560 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4460 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3899 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1110 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1080 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/7954 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4562 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4253 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1216 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -185,30 +246,55 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
+    if (mode == FileOpenMode::Read)
+        return G_SEEKABLE(g_file_read(file.get(), nullptr, nullptr));
+
+    // Embarrasingly, GIO doesn't have an atomic create-or-open function.
+    //
+    // The closest thing to create-or-open in GIO is g_file_replace() and g_file_replace_readwrite(), but they have some problems:
+    // (a) In the case of modifying an existing file, they return a new file in /tmp, doing a move swap when closing.
+    //     Even calling g_output_stream_flush() won't update the original file.
+    // (b) g_file_replace_readwrite() followed by a g_io_stream_close() will truncate the file, but we don't want ReadWrite to truncate.
+    //
+    // There is no pretty way out of this. Two naive solutions are:
+    // (a) We try to create, if it fails we try to open -> If the file is deleted between the two operations we fail.
+    // (b) We try to open, if it fails we try to create -> If the file is created between the two operations we fail.
+    //
+    // The most reliable thing we can do is spinlock retrying the other operation when we get a G_IO_ERROR_NOT_FOUND or G_IO_ERROR_EXISTS.
```
Hm... this is pretty intense, but I can't think of anything better.

g_file_replace() is normally what you want because semantically you are atomically replacing the entire file with the contents of the stream when the stream is closed. So it's a true atomic replace: nothing gets written unless the whole thing gets written. But that indeed does not work very well if you just want to create the file atomically and then leave the stream open. :/

mcatanzaro commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -185,30 +246,55 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
+        // We need to open the file as readwrite because there is no write-only open in GIO.
+        // We cannot even extract a GFileOutputStream from the GFileIOStream after the fact: the closest thing is
+        // g_io_stream_get_output_stream(), but that returns a GOutputStream which is not a GFileOutputStream and
+        // doesn't implement the GSeekable interface.
```
Did you try casting the GOutputStream* to a GFileOutputStream* to be sure it's not really a GFileOutputStream? It probably is. Would be really weird if not.

ntrrgc replied:
Hm, you're right. I must have tested something wrong. It was pretty late. Changing that...

ntrrgc replied:
Okay, not so easy: although you're right that the streams are GFileOutputStream... returning that would require unref'ing the GFileIOStream, and that closes the file, even if you ref the inner GFileOutputStream before.

ntrrgc replied:
nvm, I was getting the latter from a typo in an unrelated change.

ntrrgc replied:
nvm²: "Stream is already closed" is emitted by any write attempt, so yeah, it's closed by the container GIOStream.

mcatanzaro reviewed (commented):
Huh, I agree changing PlatformFileHandle from GFileIOStream* to GSeekable* was a good idea. (Although there is another way to represent a read-only or write-only open: get a GFileIOStream and then manually close the unwanted stream.)

But now if you were able to make this change so easily, that implies that not much of our code was actually expecting PlatformFileHandle to be a GFileIOStream*. It looks like openTemporaryFile still returns a GFileIOStream*, as does openFile when called with FileOpenMode::ReadWrite, so after your commit here, platform-specific calls to those functions are the only places where we might have code that still expects PlatformFileHandle to be a GFileIOStream*. There are probably not many such places. If we're lucky, there might be zero: I see you were able to change FileOpenMode::Read and FileOpenMode::Truncate without needing code changes elsewhere, for example.

Conclusion: we can probably delete most of FileSystemGlib.cpp and use FileSystemPOSIX.cpp instead for everything except the GLib-specific APIs (which do not use PlatformFileHandle)? Then we no longer have to maintain this alternative implementation at all. GIO APIs are usually great because they're usually simpler than using POSIX directly, but ultimately it's just a wrapper above the POSIX functions, and here they seem to be just getting in our way. (We might need to audit FileSystemPOSIX.cpp to make sure the code is good. I see a close() without any error checking, for example, which I don't like.)

Anyway, if you don't like that idea, then what you have here is probably fine. Although I'd be tempted to add error handling throughout the file, because FileSystemGlib.cpp ignores errors all over the place, and that's scary to me, but I worry callers might be expecting these functions to fail without error messages.

mcatanzaro reviewed (approved):
Anyway, I'm approving this in case you decide to proceed with this approach, and also because the tests are nice. But sounds like we both agree that deleting all of this would be even better.

ntrrgc committed:
[WTF][GLib] Rework FileSystem::openFile()
https://bugs.webkit.org/show_bug.cgi?id=254813

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

Fixes REGRESSION(259689@main). The previous implementation failed when
asked to create a file in ReadWrite mode.

This patch also increases the test coverage for FileSystem::openFile().
Previously, few argument combinations where tested, which caused the
regression to go unnoticed.

Read requests no longer create a read-write stream.

To accomodate that, PlatformFileHandle in GLib is now the generic
GSeekable* rather than GFileIOStream* (which implies a read-write file
handle).

This patch also introduces WTF::FileSystem::posixFileDescriptor() for
non-Windows platforms, which allows decoupling of the GIO API in a few
places in WebKit.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::genericGIOFileClose):
(WTF::FileSystemImpl::genericGIOFileQueryInfo):
(WTF::FileSystemImpl::genericGIOGetInputStream):
(WTF::FileSystemImpl::genericGIOGetOutputStream):
(WTF::FileSystemImpl::genericGIOGetFileDescriptorBased):
(WTF::FileSystemImpl::posixFileDescriptor):
(WTF::FileSystemImpl::fileSize):
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::openTemporaryFile):
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::closeFile):
(WTF::FileSystemImpl::seekFile):
(WTF::FileSystemImpl::truncateFile):
(WTF::FileSystemImpl::flushFile):
(WTF::FileSystemImpl::writeToFile):
(WTF::FileSystemImpl::readFromFile):
(WTF::FileSystemImpl::lockFile):
(WTF::FileSystemImpl::unlockFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::posixFileDescriptor):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/263367@main
